From 44e9b424ac7d4b7f28688da0c93ee8a4520dae14 Mon Sep 17 00:00:00 2001
From: Song Bing <bing.song@nxp.com>
Date: Mon, 24 Oct 2016 11:00:07 -0400
Subject: [PATCH 13/16] dmabuf: Add free function

Add free function to free subclass and fd memory resource.

Upstream-Status: Pending

https://bugzilla.gnome.org/show_bug.cgi?id=768794
---
 gst-libs/gst/allocators/gstdmabuf.c   | 13 +++++++++++++
 gst-libs/gst/allocators/gstdmabuf.h   |  3 +++
 gst-libs/gst/allocators/gstfdmemory.c |  6 ++++++
 gst-libs/gst/allocators/gstfdmemory.h |  9 +++++++++
 4 files changed, 31 insertions(+)

diff --git a/gst-libs/gst/allocators/gstdmabuf.c b/gst-libs/gst/allocators/gstdmabuf.c
index 7d6bcab..3c6820f 100644
--- a/gst-libs/gst/allocators/gstdmabuf.c
+++ b/gst-libs/gst/allocators/gstdmabuf.c
@@ -45,8 +45,21 @@ GST_DEBUG_CATEGORY_STATIC (dmabuf_debug);
 G_DEFINE_TYPE (GstDmaBufAllocator, gst_dmabuf_allocator, GST_TYPE_FD_ALLOCATOR);
 
 static void
+gst_dmabuf_allocator_free (GstAllocator * allocator, GstMemory * memory)
+{
+  GstDmaBufAllocatorClass *aclass;
+
+  aclass = GST_DMABUF_ALLOCATOR_GET_CLASS (allocator);
+  if (aclass->free)
+    aclass->free (allocator, memory);
+}
+
+static void
 gst_dmabuf_allocator_class_init (GstDmaBufAllocatorClass * klass)
 {
+  GstFdAllocatorClass *fd_allocator_class = GST_FD_ALLOCATOR_CLASS  (klass);
+
+  fd_allocator_class->free = gst_dmabuf_allocator_free;
 }
 
 static void
diff --git a/gst-libs/gst/allocators/gstdmabuf.h b/gst-libs/gst/allocators/gstdmabuf.h
index c6301da..166bbfe 100644
--- a/gst-libs/gst/allocators/gstdmabuf.h
+++ b/gst-libs/gst/allocators/gstdmabuf.h
@@ -90,6 +90,9 @@ struct _GstDmaBufAllocatorClass
 {
   GstFdAllocatorClass parent_class;
 
+  /*< public >*/
+  void         (*free)       (GstAllocator *allocator, GstMemory *memory);
+
   /*< private >*/
   gpointer _gst_reserved[GST_PADDING];
 };
diff --git a/gst-libs/gst/allocators/gstfdmemory.c b/gst-libs/gst/allocators/gstfdmemory.c
index ad428a7..6d36c11 100644
--- a/gst-libs/gst/allocators/gstfdmemory.c
+++ b/gst-libs/gst/allocators/gstfdmemory.c
@@ -56,6 +56,12 @@ typedef struct
 static void
 gst_fd_mem_free (GstAllocator * allocator, GstMemory * gmem)
 {
+  GstFdAllocatorClass *aclass;
+
+  aclass = GST_FD_ALLOCATOR_GET_CLASS (allocator);
+  if (aclass->free)
+    aclass->free (allocator, gmem);
+
 #ifdef HAVE_MMAP
   GstFdMemory *mem = (GstFdMemory *) gmem;
 
diff --git a/gst-libs/gst/allocators/gstfdmemory.h b/gst-libs/gst/allocators/gstfdmemory.h
index 40c90d0..93cb46b 100644
--- a/gst-libs/gst/allocators/gstfdmemory.h
+++ b/gst-libs/gst/allocators/gstfdmemory.h
@@ -69,11 +69,20 @@ typedef enum {
 struct _GstFdAllocator
 {
   GstAllocator parent;
+
+  /*< private >*/
+  gpointer _gst_reserved[GST_PADDING];
 };
 
 struct _GstFdAllocatorClass
 {
   GstAllocatorClass parent_class;
+
+  /*< public >*/
+  void         (*free)       (GstAllocator *allocator, GstMemory *memory);
+
+  /*< private >*/
+  gpointer _gst_reserved[GST_PADDING];
 };
 
 GType gst_fd_allocator_get_type (void);
-- 
1.9.1

